### PR TITLE
Improved Avro error reporting related to schemas

### PR DESCRIPTION
**Pull request by dlvenable**
### Description

This PR has two major updates:
* When starting Data Prepper, validate the provided schema. Prior to this PR it was happening only when an event is received.
* When a field does not exist in the Avro schema return an explicit exception. Prior to this PR, it was some vague message that was not possible to track down.
 
Additionally, this PR includes some code clean-up.

### Issues Resolved

N/A
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java [hunk 1]
```diff
@@ -34,9 +33,8 @@
 @DataPrepperPlugin(name = "avro", pluginType = OutputCodec.class, pluginConfigurationType = AvroOutputCodecConfig.class)
 public class AvroOutputCodec implements OutputCodec {
 
-    private static final List<String> nonComplexTypes = Arrays.asList("int", "long", "string", "float", "double", "bytes");
+    private static final List<String> NON_COMPLEX_TYPES = Arrays.asList("int", "long", "string", "float", "double", "bytes");
```
Comment (chenqi0805) [inline data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java:36]: Is PRIMITIVE_TYPES more precise here?

File: data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java [hunk 2]
```diff
@@ -179,6 +176,9 @@ private GenericRecord buildAvroRecord(final Schema schema, final Map<String, Obj
                 continue;
             }
             final Schema.Field field = schema.getField(key);
+            if(field == null) {
```
Comment (cmanning09) [inline data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java:179]: (Only because you fixed other spacing issues in this PR)

You are missing a space after the `if`

```
if (field == null) {
```

File: data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroSchemaParserFromS3.java [hunk 3]
```diff
@@ -44,7 +44,8 @@ private static String getS3SchemaObject(AvroOutputCodecConfig config) throws IOE
                 .key(config.getFileKey())
                 .build();
         ResponseInputStream<GetObjectResponse> s3Object = s3Client.getObject(getObjectRequest);
-        final Map<String, Object> stringObjectMap = objectMapper.readValue(s3Object, new TypeReference<>() {});
+        final Map<String, Object> stringObjectMap = objectMapper.readValue(s3Object, new TypeReference<>() {
+        });
```
Comment (cmanning09) [inline data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroSchemaParserFromS3.java:48]: Was this change intentional to split into two lines?

**Review by cmanning09 — APPROVED** [review]
LG - The only feedback I have are stylistic based on the code clean up you have already done.

File: data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroSchemaParserFromS3.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroSchemaParserFromS3.java:48]: This is what my IntelliJ formatter produced.

File: data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java:36]: That probably is a better name.

Comment (dlvenable) [reply data-prepper-plugins/avro-codecs/src/main/java/org/opensearch/dataprepper/plugins/codec/avro/AvroOutputCodec.java]: I renamed this.

**Comment by dlvenable** [comment]
@cmanning09 , @chenqi0805 

I had to rebase and force push due to changes from #3122. Please take another look. Thanks!

**Review by chenqi0805 — APPROVED** [review]
